### PR TITLE
NRF52 SPI master for new SPI API

### PR DESCRIPTION
**Pull request by vanwinkeljan**
NRF52 SPI Master implementation following new SPI API.
Based on legacy NRF52 SPIM driver.

### CODE REVIEW

File: drivers/spi/spi_nrf52_master.c [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+		? (SPIM_CONFIG_CPOL_ActiveLow << SPIM_CONFIG_CPOL_Pos)
+		: (SPIM_CONFIG_CPOL_ActiveHigh << SPIM_CONFIG_CPOL_Pos);
+	spim->CONFIG |=
+	    (config->operation & SPI_MODE_CPHA)
+		? (SPIM_CONFIG_CPHA_Trailing << SPIM_CONFIG_CPHA_Pos)
+		: (SPIM_CONFIG_CPHA_Leading << SPIM_CONFIG_CPHA_Pos);
+
+	if ((config->cs != NULL) || (cfg->psel.cs == CS_UNUSED)) {
+		/* Disable HW based CS */
+		spim->PSEL.CSN |= SPIM_PSEL_CSN_CONNECT_Msk;
```
Comment (tbursztyka) [inline drivers/spi/spi_nrf52_master.c]: isn't spi_context_cs_configure() missing here?

File: drivers/spi/spi_nrf52_master.c [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+	spim->INTENSET = SPIM_INTENSET_END_Msk;
+
+	spi_context_cs_control(&data->ctx, true);
+
+	transceive_current_buffer(spim, &data->ctx);
+
+	ret = spi_context_wait_for_completion(&data->ctx);
+
+	spi_context_release(&data->ctx, ret);
```
Comment (tbursztyka) [inline drivers/spi/spi_nrf52_master.c]: you could move that line before return ret, so line 225 would be move as line 222. but it's cosmetic

File: drivers/spi/spi_nrf52_master.c [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+
+	spim->TASKS_START = 1;
+	return 0;
+}
+
+static int transceive(struct spi_config *config, const struct spi_buf *tx_bufs,
+		      u32_t tx_count, struct spi_buf *rx_bufs, u32_t rx_count,
+		      bool asynchronous, struct k_poll_signal *signal)
+{
+	int ret;
```
Comment (tbursztyka) [inline drivers/spi/spi_nrf52_master.c]: usually, pre-assigned variable declaration come first, so here ret would be the last

File: arch/arm/soc/nordic_nrf5/nrf52/Kconfig.defconfig.nrf52840_QIAA [hunk 4]
```diff
@@ -14,7 +14,7 @@ config SOC
 
 config NUM_IRQS
 	int
-	default 46
+	default 48
```
Comment (tbursztyka) [inline arch/arm/soc/nordic_nrf5/nrf52/Kconfig.defconfig.nrf52840_QIAA]: you commit message does not say why you increase this number (and how it relates to spi actually)

**Review by tbursztyka — CHANGES_REQUESTED** [review]
tiny comments

**Review by galak — CHANGES_REQUESTED** [review]
Should have also introduce device tree support for SPI

File: arch/arm/soc/nordic_nrf5/include/nrf5_common.h [hunk 5]
```diff
@@ -60,6 +60,7 @@
 #define NRF52_IRQ_RTC2_IRQn                    36
 #define NRF52_IRQ_I2S_IRQn                     37
 #define NRF52_IRQ_FPU_IRQn                     38
+#define NRF52_IRQ_SPIM3_IRQn                   47
```
Comment (galak) [inline arch/arm/soc/nordic_nrf5/include/nrf5_common.h]: should come from Device Tree

File: arch/arm/soc/nordic_nrf5/nrf52/Kconfig.defconfig.nrf52840_QIAA (hunk 4 above)
Comment (vanwinkeljan) [reply arch/arm/soc/nordic_nrf5/nrf52/Kconfig.defconfig.nrf52840_QIAA]: updated commit message to indicate  number was increased as per documentation

The lower number of interrupts was causing a build error as the last SPI device has number 47

**Comment by vanwinkeljan** [comment]
updated code after testing

**Comment by vanwinkeljan** [comment]
@galak added initial attempt for dts

**Comment by vanwinkeljan** [comment]
Addresses NRF52 SPI master part of #4713

**Comment by carlescufi** [comment]
@vanwinkeljan I know @anangl is preparing to submit a driver based on `nrfx` which now lives in `ext/hal/nordic`. There's also a rework of the SPI API that @tbursztyka is leading and that will affect this. I think we should join efforts here and provide an nrfx-powered driver that complies with the new API. @anangl will be back next week from holiday and can discuss this further.

**Comment by mbolivar** [comment]
I'm trying to get this to build for nrf52_pca10040 by adding the following to hello_world's prj.conf:

```
# CONFIG_SPI_LEGACY_API is not set
CONFIG_SPI=y
CONFIG_SPI_0=y
CONFIG_SPI_NRF5=y
CONFIG_SPI0_NRF52_MASTER=y
```

I get a lot of warnings and errors: https://gist.github.com/mbolivar/13d9da829fd5d0d6501c604fc0380d29

Am I missing something?

**Comment by vanwinkeljan** [comment]
@mbolivar I will have a look at it.

For now I know there is a problem for SPI0 if I switch to SPI1 I do not seen the warnings.
But even then I get compile errors for nrf52_pca10040 due to missing declarations (I tested with nrf52840_pca10056).

One further remark is that you will also need to add a DTS overlay (nrf52_pca10040.overlay) to the hello_world directory that enables the SPI port and declares which pins it will use.
e.g.
&spi1 {
        status = "ok";
        clk-pin = <3>;
        mosi-pin = <28>;
        miso-pin = <29>;
};

**Comment by vanwinkeljan** [comment]
1. Corrected SPI Kconfig to take in account that DTS can be used for SPI0
2. Take in account that there is only a psel CSN for nrf52840

@mbolivar this should fix your build issues

**Comment by mbolivar** [comment]
@vanwinkeljan looks like the PR from @anangl (https://github.com/zephyrproject-rtos/zephyr/pull/6458) got taken in https://github.com/zephyrproject-rtos/zephyr/pull/5921 -- earlier @carlescufi mentioned this code as a potential place to work together. Any comments on where you want this PR to go? Sorry that I haven't gotten around to re-testing, as the patches in 6458 did the trick for me.

**Comment by vanwinkeljan** [comment]
@mbolivar I noticed @anangl PR and skimmed through and had no remarks.
I will close this PR

**Pull request closed without merging**